# Exercise 13

This particular Automobile Data Set includes a good mix of categorical values as well as continuous values and serves as a useful example that is relatively easy to understand. Since domain understanding is an important aspect when deciding how to encode various categorical values - this data set makes a good case study.

Read the data into Pandas

In [1]:
import pandas as pd

# Define the headers since the data does not have any
headers = ["symboling", "normalized_losses", "make", "fuel_type", "aspiration",
           "num_doors", "body_style", "drive_wheels", "engine_location",
           "wheel_base", "length", "width", "height", "curb_weight",
           "engine_type", "num_cylinders", "engine_size", "fuel_system",
           "bore", "stroke", "compression_ratio", "horsepower", "peak_rpm",
           "city_mpg", "highway_mpg", "price"]

# Read in the CSV file and convert "?" to NaN
df = pd.read_csv("http://mlr.cs.umass.edu/ml/machine-learning-databases/autos/imports-85.data",
                  header=None, names=headers, na_values="?" )
df.head()

symboling  normalized_losses         make fuel_type aspiration num_doors  \
0          3                NaN  alfa-romero       gas        std       two   
1          3                NaN  alfa-romero       gas        std       two   
2          1                NaN  alfa-romero       gas        std       two   
3          2              164.0         audi       gas        std      four   
4          2              164.0         audi       gas        std      four   

    body_style drive_wheels engine_location  wheel_base  ...  engine_size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   

   fuel_system  bore  stroke compression_ratio horsepower  peak_rpm city_mpg  \
0         mpfi  3.47    2.68               9.0      111.0    5000.0       21   
1         mpfi  3.47    2.68               9.0      111.0    5000.0       21   
2         mpfi  2.68    3.47               9.0      154.0    5000.0       19   
3         mpfi  3.19    3.40              10.0      102.0    5500.0       24   
4         mpfi  3.19    3.40               8.0      115.0    5500.0       18   

   highway_mpg    price  
0           27  13495.0  
1           27  16500.0  
2           26  16500.0  
3           30  13950.0  
4           22  17450.0  

[5 rows x 26 columns]

In [2]:
df.shape

(205, 26)

In [3]:
df.dtypes

symboling              int64
normalized_losses    float64
make                  object
fuel_type             object
aspiration            object
num_doors             object
body_style            object
drive_wheels          object
engine_location       object
wheel_base           float64
length               float64
width                float64
height               float64
curb_weight            int64
engine_type           object
num_cylinders         object
engine_size            int64
fuel_system           object
bore                 float64
stroke               float64
compression_ratio    float64
horsepower           float64
peak_rpm             float64
city_mpg               int64
highway_mpg            int64
price                float64
dtype: object

In [4]:
obj_df = df.select_dtypes(include=['object']).copy()
obj_df.head()

make fuel_type aspiration num_doors   body_style drive_wheels  \
0  alfa-romero       gas        std       two  convertible          rwd   
1  alfa-romero       gas        std       two  convertible          rwd   
2  alfa-romero       gas        std       two    hatchback          rwd   
3         audi       gas        std      four        sedan          fwd   
4         audi       gas        std      four        sedan          4wd   

  engine_location engine_type num_cylinders fuel_system  
0           front        dohc          four        mpfi  
1           front        dohc          four        mpfi  
2           front        ohcv           six        mpfi  
3           front         ohc          four        mpfi  
4           front         ohc          five        mpfi

# Exercise 13.1

Does the database contain missing values? If so, replace them using one of the methods explained in class

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
symboling            205 non-null int64
normalized_losses    164 non-null float64
make                 205 non-null object
fuel_type            205 non-null object
aspiration           205 non-null object
num_doors            203 non-null object
body_style           205 non-null object
drive_wheels         205 non-null object
engine_location      205 non-null object
wheel_base           205 non-null float64
length               205 non-null float64
width                205 non-null float64
height               205 non-null float64
curb_weight          205 non-null int64
engine_type          205 non-null object
num_cylinders        205 non-null object
engine_size          205 non-null int64
fuel_system          205 non-null object
bore                 201 non-null float64
stroke               201 non-null float64
compression_ratio    205 non-null float64
horsepower           203 non-

In [6]:
pr= pd.DataFrame(df.isnull().sum())
a=pr[pr[0] > 0]
a

0
normalized_losses  41
num_doors           2
bore                4
stroke              4
horsepower          2
peak_rpm            2
price               4

In [7]:
a=a.index.tolist()
a

['normalized_losses',
 'num_doors',
 'bore',
 'stroke',
 'horsepower',
 'peak_rpm',
 'price']

In [8]:
df[a].describe()

normalized_losses        bore      stroke  horsepower     peak_rpm  \
count         164.000000  201.000000  201.000000  203.000000   203.000000   
mean          122.000000    3.329751    3.255423  104.256158  5125.369458   
std            35.442168    0.273539    0.316717   39.714369   479.334560   
min            65.000000    2.540000    2.070000   48.000000  4150.000000   
25%            94.000000    3.150000    3.110000   70.000000  4800.000000   
50%           115.000000    3.310000    3.290000   95.000000  5200.000000   
75%           150.000000    3.590000    3.410000  116.000000  5500.000000   
max           256.000000    3.940000    4.170000  288.000000  6600.000000   

              price  
count    201.000000  
mean   13207.129353  
std     7947.066342  
min     5118.000000  
25%     7775.000000  
50%    10295.000000  
75%    16500.000000  
max    45400.000000

In [9]:
print('Media\tMediana\tModa\n',df.normalized_losses.mean(),'\t',df.normalized_losses.median(),'\t',df.normalized_losses.mode()[0])

Media	Mediana	Moda
 122.0 	 115.0 	 161.0


In [10]:
df.normalized_losses.fillna(df.normalized_losses.median(), inplace=True)

In [11]:
c=pd.DataFrame(df.horsepower.unique(),columns=['hp'])
c['rpm']=pd.DataFrame(df.peak_rpm.unique())
c

hp     rpm
0   111.0  5000.0
1   154.0  5500.0
2   102.0  5800.0
3   115.0  4250.0
4   110.0  5400.0
5   140.0  5100.0
6   160.0  4800.0
7   101.0  6000.0
8   121.0  4750.0
9   182.0  4650.0
10   48.0  4200.0
11   70.0  4350.0
12   68.0  4500.0
13   88.0  5200.0
14  145.0  4150.0
15   58.0  5600.0
16   76.0  5900.0
17   60.0  5750.0
18   86.0     NaN
19  100.0  5250.0
20   78.0  4900.0
21   90.0  4400.0
22  176.0  6600.0
23  262.0  5300.0
24  135.0     NaN
25   84.0     NaN
26   64.0     NaN
27  120.0     NaN
28   72.0     NaN
29  123.0     NaN
30  155.0     NaN
31  184.0     NaN
32  175.0     NaN
33  116.0     NaN
34   69.0     NaN
35   55.0     NaN
36   97.0     NaN
37  152.0     NaN
38  200.0     NaN
39   95.0     NaN
40  142.0     NaN
41  143.0     NaN
42  207.0     NaN
43  288.0     NaN
44    NaN     NaN
45   73.0     NaN
46   82.0     NaN
47   94.0     NaN
48   62.0     NaN
49   56.0     NaN
50  112.0     NaN
51   92.0     NaN
52  161.0     NaN
53  156.0     NaN
54   52.0     NaN
55   85.0     NaN
56  114.0     NaN
57  162.0     NaN
58  134.0     NaN
59  106.0     NaN

In [12]:
b=['horsepower',
 'peak_rpm']
for columna in b:
    df[columna].fillna(df[columna].mode()[0], inplace=True)

In [13]:
b=['bore',
 'stroke',
 'price']

In [14]:
for columna in b:
    df[columna].fillna(df[columna].mean(), inplace=True)
df[a].describe()

normalized_losses        bore      stroke  horsepower    peak_rpm  \
count         205.000000  205.000000  205.000000  205.000000   205.00000   
mean          120.600000    3.329751    3.255423  103.902439  5129.02439   
std            31.805105    0.270844    0.313597   39.680343   478.40526   
min            65.000000    2.540000    2.070000   48.000000  4150.00000   
25%           101.000000    3.150000    3.110000   70.000000  4800.00000   
50%           115.000000    3.310000    3.290000   95.000000  5200.00000   
75%           137.000000    3.580000    3.410000  116.000000  5500.00000   
max           256.000000    3.940000    4.170000  288.000000  6600.00000   

              price  
count    205.000000  
mean   13207.129353  
std     7868.768212  
min     5118.000000  
25%     7788.000000  
50%    10595.000000  
75%    16500.000000  
max    45400.000000

# Exercise 13.2

Split the data into training and testing sets

Train a Random Forest Regressor to predict the price of a car using the nominal features

In [15]:
#Import the libraries:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [16]:
#Define the data:
nom_df = df.select_dtypes(include=['int64','float64']).copy()
nom_df.head()

symboling  normalized_losses  wheel_base  length  width  height  \
0          3              115.0        88.6   168.8   64.1    48.8   
1          3              115.0        88.6   168.8   64.1    48.8   
2          1              115.0        94.5   171.2   65.5    52.4   
3          2              164.0        99.8   176.6   66.2    54.3   
4          2              164.0        99.4   176.6   66.4    54.3   

   curb_weight  engine_size  bore  stroke  compression_ratio  horsepower  \
0         2548          130  3.47    2.68                9.0       111.0   
1         2548          130  3.47    2.68                9.0       111.0   
2         2823          152  2.68    3.47                9.0       154.0   
3         2337          109  3.19    3.40               10.0       102.0   
4         2824          136  3.19    3.40                8.0       115.0   

   peak_rpm  city_mpg  highway_mpg    price  
0    5000.0        21           27  13495.0  
1    5000.0        21           27  16500.0  
2    5000.0        19           26  16500.0  
3    5500.0        24           30  13950.0  
4    5500.0        18           22  17450.0

In [17]:
len(nom_df.columns)

16

In [18]:
#Define the features and the output:
X = nom_df.drop('price',axis=1)

In [19]:
y = nom_df['price']

In [20]:
# Split the data:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
n_samples = X_train.shape[0]

**Random Forest**

In [21]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import cross_val_score

In [22]:
pd.Series(np.sqrt(-cross_val_score(RandomForestRegressor(n_estimators=100), X, y, cv=10, scoring='neg_mean_squared_error'))).describe()

count      10.000000
mean     3307.545249
std      1573.889552
min       896.217753
25%      2153.766229
50%      2936.070045
75%      4722.338861
max      5501.213848
dtype: float64

# Exercise 13.3

Create dummy variables for the categorical features

Train a Random Forest Regressor and compare

In [23]:
#Redifining X
X = df.drop('price',axis=1)

In [24]:
#Get Dummies
X = pd.get_dummies(X)
X.head()

symboling  normalized_losses  wheel_base  length  width  height  \
0          3              115.0        88.6   168.8   64.1    48.8   
1          3              115.0        88.6   168.8   64.1    48.8   
2          1              115.0        94.5   171.2   65.5    52.4   
3          2              164.0        99.8   176.6   66.2    54.3   
4          2              164.0        99.4   176.6   66.4    54.3   

   curb_weight  engine_size  bore  stroke  ...  num_cylinders_twelve  \
0         2548          130  3.47    2.68  ...                     0   
1         2548          130  3.47    2.68  ...                     0   
2         2823          152  2.68    3.47  ...                     0   
3         2337          109  3.19    3.40  ...                     0   
4         2824          136  3.19    3.40  ...                     0   

   num_cylinders_two  fuel_system_1bbl  fuel_system_2bbl  fuel_system_4bbl  \
0                  0                 0                 0                 0   
1                  0                 0                 0                 0   
2                  0                 0                 0                 0   
3                  0                 0                 0                 0   
4                  0                 0                 0                 0   

   fuel_system_idi  fuel_system_mfi  fuel_system_mpfi  fuel_system_spdi  \
0                0                0                 1                 0   
1                0                0                 1                 0   
2                0                0                 1                 0   
3                0                0                 1                 0   
4                0                0                 1                 0   

   fuel_system_spfi  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  

[5 rows x 75 columns]

In [25]:
# Split the data:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
n_samples = X_train.shape[0]

In [26]:
pd.Series(np.sqrt(-cross_val_score(RandomForestRegressor(n_estimators=100), X, y, cv=10, scoring='neg_mean_squared_error'))).describe()

count      10.000000
mean     3261.775247
std      1594.361642
min       755.524181
25%      2215.646475
50%      2956.222527
75%      4475.873549
max      5534.799832
dtype: float64

This model has a less min RMSE (755.52) compared with the model without the categorical features (896.21)

# Exercise 13.4

Apply two other methods of categorical encoding

compare the results

In [27]:
import category_encoders as ce

**Polinomial coding**

In [28]:
X = ce.PolynomialEncoder().fit_transform(df.drop('price',axis=1))

In [29]:
# Split the data:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
n_samples = X_train.shape[0]

In [30]:
pd.Series(np.sqrt(-cross_val_score(RandomForestRegressor(n_estimators=100), X, y, cv=10, scoring='neg_mean_squared_error'))).describe()

count      10.000000
mean     3315.140387
std      1459.939712
min      1846.158883
25%      2120.049055
50%      2895.529784
75%      4396.195778
max      5600.940092
dtype: float64

**Helmert coding**

In [31]:
X = ce.HelmertEncoder().fit_transform(df.drop('price',axis=1))

In [32]:
# Split the data:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
n_samples = X_train.shape[0]

In [33]:
pd.Series(np.sqrt(-cross_val_score(RandomForestRegressor(n_estimators=100), X, y, cv=10, scoring='neg_mean_squared_error'))).describe()

count      10.000000
mean     3257.392712
std      1680.584603
min       768.381869
25%      2136.466962
50%      2977.648853
75%      4364.998325
max      6109.525212
dtype: float64

**Conclusion:**  
If we compare all the encoding used, we can conslude that the best RMSE (compared the min RMSE), can be obtained from the dummy variables (755.52), the second with the Helmert coding (768.38), third without the categorical features (896.21), and the worst encoding, for this case, is the Polinomial Coding.